Publish minutes of 2025-11-10 - 2025-11-14 TPAC meetings (#923)
hober/web-extensions: _minutes/2025-11-10-wecg-tpac.md

(all times JST, UTC+9)

 * Welcome – 11:00 to 11:30
 * Working Group Charter – 11:30 to 12:30
 * (Lunch – 12:30 to 13:45)
 * WPT Testing Status – 13:45 to 14:15
 * User Attention – 14:15 to 15:00
 * (Break – 15:00 to 15:30)
 * Extension Origin – 15:30 to 16:00
 * Bug Triage – 16:00 to 16:45

## Attendees

 * Rob Wu (Mozilla)
 * Simeon Vincent (Mozilla)
 * Kiara Rose (Apple)
 * Devlin Cronin (Google Chrome)
 * Tomislav Jovanovic (Mozilla)
 * Carlos Jeurissen (Independent)
 * Oliver Dunk (Google Chrome)
 * Felipe Erias (Igalia)
 * Kanaru Sato (...)
 * Ehsan Toreini (Samsung)
 * Vadim Makeev (MDN @ Mozilla)
 * Christian Brand (1Password)
 * Rene Leveille (1Password)
 * Mike (sideshowbarker) (W3C)
 * Andrew Dekker (Meta)
 * Ricky Mondello (Apple)
 * Takagi (KDDI)
 * Timothy Hatcher (Apple)
 * Mukul Purohit (Microsoft)
 * Casey Garland (Capital One)
 * George Henderson (Capital One)
 * Binh Trong Hoang (Huawei)
 * Samuel Maddock (Salesforce)
 * Sarven Capadisli (TAG)

## Notes

### Welcome

11:00-11:30
Introductions, goals, and overview of the week's objectives.

 * [simeon] Intro; overview of today's agenda.
 * [round of intros]

>  * [simeon] Objectives: Creation of WG; wpt,

Reviewer: ```suggestion
 * [simeon] Objectives: Creation of WG; WPT
```

   * [mike] Understand the desire for non-interoperability. Could add a link from the charter to the other explanatory part.
   * [devlin] The quote above from the WECG, would like to include that in the WEWG charter.
   * [timothy] Do we expect to have many things that are experimental landing in the working group? Doing that in the community group makes sense.
   * [devlin] Any spec we produce should be aligned between browser. But something we don't specify is literally everything that falls under the browser namespace, browsers may add to the browser namespace that is not specified.
   * [mike] The CG can have a wider community, doesn't need to sign up with patent policy etc. The purpose of the WG is to publish something stable, once it reaches some level of maturity.
   * [devlin] I expect the majority of the discussions to continue in the CG. The spec published by the WG would not exhaustively list the full capabilities of extensions; implementing more would not fail a browser to be spec-compliant.
   * [mike] One you write something down, including examples would be useful.

>    * [simeon] system.\* APIs.

Reviewer: ```suggestion
   * [simeon] `system.*` APIs.
```

   * [devlin] I expect we'd get pushback if we tried to put the scope in a separate document. Having a doc that explains why the spec is non-exhaustive.
   * [mike] Like a FAQ. Don't want to answer the same question a hundred times from different people.
 * [rob] Next steps?
   * [devlin] Would be nice to resolve the feedback threads (in Github), update the charter and invite csarven (TAG) here at TPAC to discuss the next steps.
   * Action Item: Simeon to resolve comment threads, in addition to adding blurb from CG about non-spec'd APIs
   * [simeon] Once we have the revision, what are the next steps in the W3?
   * [mike] Next step is to formally send it out to the advisory committee (AC). Other companies can see it. Patent disclosures can happen upon joining groups, royalty-free licensing is needed to publish specs that may infringe on specific patents. Related, disclosure obligations: even if not part of a WG, if a company is aware of relevant patents, companies have an obligation to mention relevant parts. Sending to the AC is an opportunity to get this review and feedback.

>    * [mike] A problem might arise if we don't get enough responses from companies. E.g. if there are 400 companies, and not 40 (10%) responses..

Reviewer: ```suggestion
   * [mike] A problem might arise if we don't get enough responses from companies. E.g. if there are 400 companies, and not 40 (10%) responses.
```

   * [rob] Is it a problem to create the WG? For whom is it a problem?
   * [mike] Problem for me and W3C. I would have to recommend advancing the group despite not meeting the threshold. Strongly recommend you make sure your own organizations respond to the AC announcement. Filing in the form can be completed in 5 minutes.
   * [rob] Does that undermine the legitimacy of the group or block its formation?
   * [mike] Responses are a strong signal that we have sufficient participation from members that think we should actually be spending time and energy on a given area of work.
   * [rob] Sounds like next steps are submitting it to the AC. Anything else?
   * [mike] There are some other minor details, but they're mostly housekeeping. Once you've got the charter in a state you're happy with, there's not much else to do for a while.

(Lunch 12:30 - 13:45)

### Web Platform Tests Status

13:45 - 14:15
Review the current state of Web Platform Tests for extensions. Discuss next steps for expanding test coverage and whether AI tools could help convert existing tests into WPT format.

>  * [simeon] Start with review. Where are we with wpt integration?

Reviewer: ```suggestion
 * [simeon] Start with review. Where are we with WPT integration?
```

 * [kiara] Landed. We have 1 test. Should definitely get more coverage.
 * [rob] As in everything is running in Safari/WebKit?
 * [timothy] Not sure how we integrate with the WPT test suite. Safari Tech Preview (STP) integration should work. Don't know if that's running on wpt.fyi today.
 * [rob] But you are able to run them locally during development.
 * [kiara] Should be testable against Safari as well as STP.
 * [tomislav] Also have all the code landed. Haven't confirmed whether they run outside our infrastructure (on wpt.fyi). That's my next task. May need to expose more detailed results. Also want to write several more tests of the test framework to confirm/demonstrate how tests work. Will be useful for Google when they are ready to begin implementation.
 * [rob] Is the syncing of tests already set up?
 * [tomislav] Yes.
 * [rob] If someone were to write the tests manually, are they able to run them?
 * [tomislav] Yes. Should be runnable using standard WPT infrastructure.
 * [kiara] tests results are being reported on wpt.fyi : https://wpt.fyi/results/web-extensions?label=experimental&label=master&aligned

>  * [tomislav] The tests also run on our infrastructure. If you add a test, the bot will pick up the change and sync it with WPT. Carlos has created a patch that adds runtime.getVersion() support, which would be a good candidate for the wpt test structure.

Reviewer: ```suggestion
 * [tomislav] The tests also run on our infrastructure. If you add a test, the bot will pick up the change and sync it with WPT. Carlos has created a patch that adds runtime.getVersion() support, which would be a good candidate for the WPT test structure.
```

 * [george] I feel like there's enough there to explore.
 * [rob] I think we can close this issue as an action item.
 * [devlin] Agree, I think we can say we understand the use cases and are interested in exploring suggestions on a case by case basis.
   * **Action Item**: Devlin to close issue
   * [Done](https://github.com/w3c/webextensions/issues/892#issuecomment-3509545887)
 * [george] We should talk about declarative badging, but there might not be enough time to cover it.
 * [devlin] Might be, not much has changed. Browsers are generally supportive but no one has had time to explore approaches or write up a proposal.

### Extension Origin

15:30 - 16:00
Discuss how to define a stable and predictable origin for extensions in cases where one is required, and related privacy or security considerations such as missing Referer headers in extension iframes.
See [issue 238](https://github.com/w3c/webextensions/issues/238#issuecomment-3470364772) and [issue 896](https://github.com/w3c/webextensions/issues/896) for background.

>  * [rob] The latest motivation for this was the fact that the Referer header is omitted, which can be a signal that the request is coming from an extension. Another use case: authentication. Chrome includes the extension origin based on the extension id. We may be able to solve this generally.

Reviewer: ```suggestion
 * [rob] The latest motivation for this was the fact that the Referer header is omitted, which can be a signal that the request is coming from an extension. Another use case: authentication. Chrome includes the extension origin based on the extension ID. We may be able to solve this generally.
```

 * [carlos] Note that DNR is not an ideal solution as the extension would have to request host access. Referrer policy on iframes may help
 * [rob] Referer could be useful, but
 * [casey] Would we be able to trust that requests with that ID are only coming from our extension?
 * [tomislav] If its coming from the browser, yes.
 * [oliver] But another extension could have a DNR rule to set the header.
 * [tomislav] Right.
 * [rob] If we were to agree on a standard extension origin format, we could align on and agree that other origins wouldn't be able to spoof that header.
 * [george] How does Firefox handle origins?
 * [rob] Currently the origin is sometimes stripped and sometimes included. It changed a couple of times in specific scenarios to repair backwards compatibility.
 * [george] Should we consider how this plays into web_accessible_resources as well?
 * [simeon] Yes.
 * [tomislav] Also relevant for identity API. We generate a fake origin for the extension to use.

>  *

Reviewer: ```suggestion
```

 * [rob] A potential solution is that extensions could expose an origin based on the extension's internal concept of the extension ID and this would require an opt-in from the extension.
 * [devlin] Need to be careful with that. Origin header is only omitted on the web if the request is from the same origin.

>  * [rob]

Reviewer: @Rob--W, please either add a comment here or remove the line.

```suggestion
```

 * [oliver] How is this related to the frame embedding and webuathn cases?
 * [rob] Both specify where the request came from. I'm proposing that we define a method of having a fixed origin that's static and common to all users. Since this kind of static value can be a privacy concern, it would need to be an opt in.

>  * [oliver] Couldn't we say that the static ID is used in certain cases and the dynamic ID is used in others. Couldn't you use the bss.gecko.id value in those requests?

Reviewer: ```suggestion
 * [oliver] Couldn't we say that the static ID is used in certain cases and the dynamic ID is used in others. Couldn't you use the `browser_specific_settings.gecko.id` value in those requests?
```

 * [oliver] Are we comfortable saying extension origins can be fixed?
 * [rob] Yes, with the condition that it should be possible to make it such that a website cannot detect the presence of the extension. Any use of web_accessible_resources currently exposes the presence of the extension.
 * [tomislav] Right, in Chrome's MV3 implementation they have an opt out.
 * [rob] That does not solve it generally for extensions that want to expose something on potentially any website, but is a task for later. Right now I'd like to specify a standard origin derived from an extension identifier, with the principle that the origin cannot easily leak to websites.
 * [timothy] That's essentially our stance as well. We randomize the host name and take steps to prevent a site from detecting an extension. If you're accessing href attributes in DOM, we'll mask the URL if you're accessing it from the main world. We also mask it from error stacks. CSP reports also have masked urls.
 * [tomislav] Would you be open to aligning on fixed urls?
 * [timothy] Hard to say. If fixed origin could get out in some way and be used to detect the presence of the extension, that would be a concern.

>  * [rob] Currently extensions have to opt into use_dynamic_url in WAR, but that means that extensions that don't set it expose users to fingerprinting risks. I'm suggesting we change the default.

Reviewer: ```suggestion
 * [rob] Currently extensions have to opt into `use_dynamic_url` in WAR, but that means that extensions that don't set it expose users to fingerprinting risks. I'm suggesting we change the default.
```

   * [simeon] In the past we've used labels to flag items to resolve in the next manifest version bump, but then we have to keep the issue open. Might be best to create a new issue to collect items we'd like to address in the next manifest version bump, cross-link the issue, and close out the origin.
   * [carlos] I like that, I can create the issue.
 * [Issue 878](https://github.com/w3c/webextensions/issues/878): Proposal: `runtime.getVersion()` method
   * [carlos] Back and forth between Oliver and me. Firefox and Safari return the manifest value, whereas Chrome returns the parsed interpretation such as zeroes removed. Concerned about extensions bugs caused by this difference.
   * [oliver] The implementations are different, but I don't think that it is a meaningfully different way.
   * [timothy] Matching yourself is the critical part here.
   * [devlin] This returns the version as parsed by the browser. Use `runtime.getManifest()` if the objective is the manifest value, use `runtime.getVersion()` for browser's implementation.
   * [carlos] That removes one of the advantages of having a dedicated method.

>    * [devlin] That was one of the advantages of this approach, it's not just a shorthand for getManifest().version, it does something different from that method and consistent with other parts of the API.

Reviewer: ```suggestion
   * [devlin] That was one of the advantages of this approach, it's not just a shorthand for `getManifest().version`, it does something different from that method and consistent with other parts of the API.
```

   * [rob] Although the behaviors are different, that is not a bug in the `getVersion()` API; the returned values match the browser's interpretation.
 * [Issue 897](https://github.com/w3c/webextensions/issues/897): Add support for `tabs.create({ incognito: true })`
   * [carlos] Google folks, what do you think?
   * [devlin] No. Can't mix incognito and on-the-record tabs in the same window.

>    * [carlos] windows API isn't necessarily defined on mobile. There are already situations where (??) isn't necessarily true.

Reviewer: ```suggestion
   * [carlos] windows API isn't necessarily defined on mobile. There are already situations where … isn't necessarily true.
```

   * [rob] Idea here is not to create an incognito tab in a normal window, it's to open an incognito tab in the appropriate place.
   * [tomislav] Seems too much to overload.
   * [devlin] Agree, every magical. We use the current window if the window isn't specified. This complicates that logic.
   * [timothy] I don't see an issue with it; could search for window and create one if needed, but we already know that, switch over to the right window and create a tab.
   * [devlin] That's specific to Safari. We don't have that pairing behavior.
   * [tomislav] Right, which incognito window do you go to?
   * [rob] What happens if you have an extension that doesn't use incognito split?

>    * [devlin] If it's not split, tabs.create() goes to the most recently active window, whether that's incognito or on the record.

Reviewer: ```suggestion
   * [devlin] If it's not split, `tabs.create()` goes to the most recently active window, whether that's incognito or on the record.
```

   * [rob] Isn't that magic too, by the same definition?
   * [devlin] More magic isn't the answer.
   * [rob] The current incognito-window tie is due to Chrome's tight coupling of incognito to window definition. We see that this model does not always apply, e.g. in Safari private/non-private in the same window, and in Firefox for Android there are no windows. Offering the ability to open a tab without knowing the window could be useful for scenarios where the dev does not care about windows, but just want to open a window.
   * [oliver] Can't you open an incognito window even if your extension isn't allowed to run in incognito?
   * [simeon] I see that as similar to clicking a link in another application. Devlin, can't we have a little magic as a treat?

>    * [devlin] Would be less concerned if it didn't mean having to change the documentation around omitting the ID to tabs.open() docs.

Reviewer: ```suggestion
   * [devlin] Would be less concerned if it didn't mean having to change the documentation around omitting the ID to `tabs.open()` docs.
```

   * [rob] “tab in an incognito context”
   * [oliver] I'm okay with neutral if Devlin is okay with it.
   * [devlin] I'm okay with neutral.
   * [oliver] Willing to accept patch?
   * [devlin] Yes.
   * **Resolution**: Chrome neutral on `tabs.create({ incognito: true })` (Firefox and Safari were already supportive)
 * [Issue 903](https://github.com/w3c/webextensions/issues/903): Proposal: `browser.action.associatedTabId`
   * [carlos] This was briefly discussed in TPAC Seville. Active tab is not necessarily associated with the current tab. I believe in Spain we saw that Firefox allows users to switch the active tab while the action popup is open.
   * [oliver] Could we do something with the getContext API? Maybe something like querying currentContext: true?
   * [timothy] Sounds a bit convoluted.
   * [devlin] Does not help, because the context would be the popup.
   * [oliver] would be a new property like associatedTabId.
   * [devlin] Might be relevant for sidepanels?
   * [rob] Can we use the magic tab ID for the current popup? (examples of prior art: TAB_ID_NONE, WINDOW_ID_CURRENT)

>    * [devlin] So it's a supplement to querying for the current tab id?

Reviewer: ```suggestion
   * [devlin] So it's a supplement to querying for the current tab ID?
```

   * [rob] Yes.
   * [timothy] Devlin, what do you think of more magic?
   * [devlin] I like that this doesn't introduce a new API and is expandable to other situations. It is also not magic behavior, but a fixed constant. Carlos, one of the challenges with associatedTabId is that we don't have too many situations where a property only exists in certain contexts.
   * [carlos] I'm also okay with the magic values.
   * [oliver] I like things you can easily discover by looking through the TOC of the API docs. Arguably you could also do this with a magic property.

>    * [carlos] It's also limited by feature set. You don't know the actual value. But I guess you could call tabs.get() with it to get the actual value.

Reviewer: ```suggestion
   * [carlos] It's also limited by feature set. You don't know the actual value. But I guess you could call `tabs.get()` with it to get the actual value.
```

   * [devlin] I prefer a function over a property to get this value.

>    * [oliver] Difference with location.href getter?

Reviewer: ```suggestion
   * [oliver] Difference with `location.href` getter?
```

>    * [devlin] window.location.href tells you something about your current document. window.top or window.parent tell you something about your frame tree. The difference I see is that it tells you something about a different context. We also have to

Reviewer: ```suggestion
   * [devlin] `window.location.href` tells you something about your current document. `window.top` or `window.parent` tell you something about your frame tree. The difference I see is that it tells you something about a different context. We also have to…
```

   * [george] We'd also want to expose this to the sidepanel API.
   * [carlos] Firefox, what do you think about the magic constants for this?
   * [tomislav] +1
   * [rob] \*thumbs up\*

>    * [] browser.tabs.TAB_ID_ASSOCIATED

Reviewer: ```suggestion
   * [_unknown_] `browser.tabs.TAB_ID_ASSOCIATED`?
```
